files/metrics-collector.sh

```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env bash
+# vim: ft=sh
+set -e
+
+METRICS_DIR="${METRICS_DIR:-/var/lib/nimbus-benchmark-metrics}"
+METRICS_FILE="${METRICS_DIR}/benchmark_metrics.prom"
+METRICS_TMP="${METRICS_DIR}/.benchmark_metrics.tmp"
+
+function initMetricsDir() {
+    mkdir -p "${METRICS_DIR}"
+    if [[ ! -f "${METRICS_FILE}" ]]; then
+        touch "${METRICS_FILE}"
+    fi
+}
+
+function recordStageStart() {
+    local stage_name="$1"
+    local timestamp=$(date +%s%N)
+
+    initMetricsDir
+
+    echo "${stage_name}_start=${timestamp}" >> "${METRICS_DIR}/.stage_timings"
+}
+
+function recordStageEnd() {
+    local stage_name="$1"
+    local success="$2"  # 0 for failure, 1 for success
+    local timestamp=$(date +%s%N)
+
+    initMetricsDir
+
+    local start_time=$(grep "^${stage_name}_start=" "${METRICS_DIR}/.stage_timings" 2>/dev/null | cut -d= -f2 | tail -1)
+
+    if [[ -n "${start_time}" ]]; then
+        local duration_ns=$((timestamp - start_time))
+        local duration_seconds=$(echo "scale=6; ${duration_ns} / 1000000000" | bc)
+
+        {
+            # Filter out old metrics
+            if [[ -f "${METRICS_FILE}" ]]; then
+                grep -v "^nimbus_benchmark_stage_success{.*stage_name=\"${stage_name}\"" "${METRICS_FILE}" 2>/dev/null | \
+                grep -v "^nimbus_benchmark_stage_duration_seconds{.*stage_name=\"${stage_name}\"" || true
+            fi
+            echo "nimbus_benchmark_stage_success{stage_name=\"${stage_name}\",benchmark_type=\"${BENCHMARKING_TYPE}\"} ${success}"
+            echo "nimbus_benchmark_stage_duration_seconds{stage_name=\"${stage_name}\",benchmark_type=\"${BENCHMARKING_TYPE}\"} ${duration_seconds}"
```

Comment:
Couldn't we just do this with one `sed` command?

Reply:
oh right :D